### ISSUE TITLE
Solver Performance: JIT Compilation Overhead Dominates for Small Datasets

### ISSUE THREAD

**Issue by BalzaniEdoardo** [flatironinstitute/nemos#547]
## Summary

NeMoS's JAX-based solvers recompile on every `fit()` call. For small-to-medium datasets — which are common in neuroscience even at millisecond binning — this compilation cost dominates wall time and makes cross-validation grids prohibitively slow compared to scipy-based alternatives. Two complementary solutions are proposed.

---

### Observation 1: compilation cost outweighs compute for small datasets

JAX JIT compiles the solver at the first `fit()` call for a given abstract input signature (shapes, dtypes, pytree structure). For large models — many neurons, long recordings, complex basis expansions — this one-time cost is negligible relative to the actual optimization. For small models the ratio reverses: compilation dominates.

### Observation 2: cross-validation multiplies the cost

Sklearn's `GridSearchCV` and `cross_val_score` call `clone()` on the estimator before each fold. `clone()` reconstructs the model via `__init__`, producing a fresh instance with no compiled solver. Every fold of every grid point therefore triggers a cold JIT compilation. For a 5-fold CV over a 3-point parameter grid the total compilation count is 15.

The measured cost on a small GLM (100 samples, 2 features, 3 regularization strengths, 5-fold CV, 3 repetitions):

```
JAX L-BFGS (NeMoS default):  [4.91s, 4.85s, 4.83s]
Scipy L-BFGS-B (custom adapter from docs):  [0.074s, 0.068s, 0.067s]
```

~70x slower. For reference, the scipy adapter is a user-space ~50-line class using the existing custom solver interface documented in NeMoS.

### Root cause

Solvers are re-instantiated on every `fit()` call. This is by design: it guarantees that the solver always reflects the exact model configuration and provided kwargs. The correctness guarantee is very valuable (the current implementation is very safe) but the performance cost is very undesirable.

JAX's JIT cache is keyed on **function object identity**. A fresh solver instance creates new JIT-wrapped function objects; JAX has no knowledge that they are semantically identical to those built in the previous `fit()` call and recompiles from scratch. Two calls to `jax.jit(f)` on the same `f` produce two independent caches. The module-level solver cache fixes this by returning the same Python object for the same configuration — same object identity → same JIT cache → hit.

---

## Proposed Solutions

Both solutions are worth pursuing independently and are complementary.

### Solution 1 — Native scipy solver (simpler, no compilation)

Promote the scipy adapter pattern from a documentation example to a built-in, registered solver.

**What this means:**
- `ScipySolverBase` ships in `nemos.solvers`, implementing the full `SolverProtocol`
- `"LBFGS-scipy"` (and others: `"BFGS"`, `"Newton-CG"`, `"TNC"`) registered in the solver registry out of the box
- Declared compatible with the relevant regularizers
- A scipy solver becomes the recommended default for small-to-medium CPU-resident problems

**Why it works:** scipy optimizers are backed by compiled Fortran/C (LAPACK, BLAS). No JAX tracing overhead. Shape changes do not trigger extra computation — scipy just sees a new numpy array.

**Open design question:** scipy's L-BFGS-B supports per-parameter box constraints (`bounds`). How should users provide these for pytree-structured parameters (e.g., a GLM where weights and intercept are separate arrays)? Options:
- A `bounds` argument that mirrors the pytree structure of `params` (probably best).
- A flat `(lower, upper)` array whose length equals the total number of parameters
- A callable `bounds_fn(params) -> (lower, upper)`

**Limitations:** does not benefit from GPU/TPU acceleration. For very large population models, pure JAX solvers will still be faster once compilation is amortized.

---

### Solution 2 — Module-level solver cache (complex, JAX scalability with minimal recompilation)

Cache compiled solver instances at module level, keyed by model configuration. A cache hit reuses the compiled solver across `clone()` calls, CV folds, and repeated fits.

**Cache key** (method of `BaseRegressor`, must be overridden in subclasses that add configuration):
```python
(
    model_class,
    regularizer_class,
    observation_model_class,
    solver_algo_name,
    solver_backend,
    tuple(sorted(solver_kwargs.items())),
)
```
No data-dependent quantities (shapes, dtypes, pytree structure) in the key. JAX's own tracing handles recompilation when those change.

**Cache storage:**
```python
_SOLVER_CACHE: dict[tuple, SolverInstance] = {}
```
A frozen-dict type can be used as the value container for additional safety.

**Making the cache reflect configuration exactly — stateless regularizers and observation models**

Currently, quantities like `regularizer_strength`, the GroupLasso `mask`, and the NegativeBinomial `scale` are captured in the solver closure at construction time. This means:
- The cache cannot be reused when strength or scale changes (different fold in a grid search).
- If the pytree structure of X changes between fits (e.g., dict vs. array), the captured mask is wrong.

Fix: expose these as runtime arguments rather than closure state.

The solver interface already threads extra args through `run(self, init_params, *args)`. The proposal is to relax this to `**kwargs` at the Python level (the jitted inner functions retain explicit named parameters — JAX cannot trace `**kwargs`):

```python
# Python-level solver method
def run(self, init_params, X, y, **extra):
    return self._jitted_run(init_params, X, y, **extra)

# Jitted function — explicit named params required
@jax.jit
def _run(params, X, y, regularizer_strength, mask=None, scale=None): ...
```

`BaseRegressor` provides a base extra-args packer; subclasses extend it:
```python
# BaseRegressor — common to all models
def _get_solver_extra_kwargs(self):
    return {"regularizer_strength": self.regularizer_strength}

# NB subclass — adds its own
def _get_solver_extra_kwargs(self):
    return {**super()._get_solver_extra_kwargs(), "scale": self._scale}
```

The GroupLasso mask is computed during input preprocessing and appended by the model — it is never captured in a closure, so JAX can retrace freely when the mask structure changes.

`regularizer.penalized_loss` and `regularizer.proximal_operator` must be updated to accept `regularizer_strength` (and mask, for GroupLasso) as explicit arguments rather than closure-captured values. Same for the NegativeBinomial log-likelihood and the `scale` parameter.

**Cache invalidation:** none needed. The key encodes the full configuration. A new configuration produces a new key and a new cache entry. Old entries are never invalidated.

**Interaction with `clone()`:** `clone()` produces a model with the same class and hyperparameters → same cache key → no. jit compilation required. CV grids compile once per unique configuration, not once per fold.

#### Open questions

- Should be always enable caching or should caching be a **global config flag** (`nmo.config.enable_solver_cache = True/False`), following the `jax.config` pattern? Note that a per-instance flag would be wiped by `clone()`, defeating the purpose entirely. The global flag preserves the original safe behavior and allows clean opt-out in tests or debugging contexts.
- Future-proof safety: `_solver_cache_key()` should be **abstract** in `BaseRegressor`, forcing every concrete subclass to implement it. The base class provides a default implementation that introspects `__init__` parameters, attempts to hash each value, and **raises immediately for non-hashable parameters** with a clear message: the developer must either make the value hashable or explicitly list it in a `_cache_key_excluded_params` class attribute. This fail-loud default prevents silent cache key collisions when new attributes are added. A test iterates all concrete subclasses and checks that every `__init__` parameter is either present in the key or explicitly excluded.

---

## Project todos

[ ] Set up a benchmarking script. The script runs periodically (cron job on workstation is the simplest) and stores the output in a csv file. The benchmarking runs over different solvers, input size, input structure, gpu or cpu backends. csv should include the nemos commit hash, the hash should be used to check that any release, (tagged commit) has been benchmarked. csv must be saved in the workstation folder that is publicly accessible on the web.  Initially, use the scipy wrapper in the docs as part of the benchmarking: immediately informative on the cost of jit-compilation.

[ ] Change the doc build so that it downloads the latest csv and create a benchmarking page with plots that the user can refer to when choosing a configuration. 

[ ] Change deploy action to enforce that the csv with the benchmark corresponding to the commit was used to generate the docs for the release branch. If not, prevent the deployment.

[ ] Work on option 1: `scipy` solvers. Create a `scipy` adapter as part of nemos, wrapping `scipy.optimize.minimize`. All methods available through that interface (L-BFGS-B, TNC, SLSQP, trust-ncg, …) become available immediately at no extra cost. Register the most relevant ones by default; declare compatibility with regularizers. Update the custom solver how-to guide to use a solver from a genuinely external ecosystem (e.g., a `torch.optim` adapter or a handwritten derivative-free method) to demonstrate interface extensibility rather than a scipy method that will by then be canonical NeMoS.

[ ] Work on option 2: caching mechanism. 

---

## Priority and Scope

| Solution | Complexity | Benefit | When                                    |
|---|---|---|-----------------------------------------|
 | Benchmarking | Low | Allow informed decision on how to proceed and informs users on how to choose solvers. | Now                                     | 
| First-class scipy solver | Low | Immediate fix for small-problem CV | Near term                               |
| Module-level solver cache | High | Scalable, minimal compilation for all problem sizes | Requires regularizer interface refactor |

The scipy solver can be added without touching the regularizer interface. The caching system requires the stateless-regularizer refactor as a prerequisite.
